tests/GraphQL.Client.Tests.Common/FluentAssertions.Reactive/FluentTestObserver.cs

```diff
@@ -0,0 +1,88 @@
+using System.Diagnostics;
+using System.Reactive;
+using System.Reactive.Concurrency;
+using System.Reactive.Linq;
+using System.Threading;
+using Microsoft.Reactive.Testing;
+
+namespace GraphQL.Client.Tests.Common.FluentAssertions.Reactive
+{
+    /// <summary>
+    /// Observer for testing <see cref="Observable"/>s using the FluentAssertions framework
+    /// </summary>
+    /// <typeparam name="TPayload"></typeparam>
+    public class FluentTestObserver<TPayload> : IObserver<TPayload>, IDisposable
+    {
+        private readonly IDisposable _subscription;
+        private readonly EventLoopScheduler _observeScheduler = new EventLoopScheduler();
+        private readonly RollingReplaySubject<Recorded<Notification<TPayload>>> _rollingReplaySubject = new RollingReplaySubject<Recorded<Notification<TPayload>>>();
+
+        /// <summary>
+        /// The observable which is observed by this instance
+        /// </summary>
+        public IObservable<TPayload> Subject { get; }
+
+        /// <summary>
+        /// The stream of recorded <see cref="Notification{T}"/>s
+        /// </summary>
+        public IObservable<Recorded<Notification<TPayload>>> RecordedNotificationStream => _rollingReplaySubject.AsObservable();
+
+        /// <summary>
+        /// The recorded <see cref="Notification{T}"/>s
+        /// </summary>
+        public IEnumerable<Recorded<Notification<TPayload>>> RecordedNotifications =>
+            _rollingReplaySubject.Snapshot();
+
+        /// <summary>
+        /// The recorded messages
+        /// </summary>
+        public IEnumerable<TPayload> RecordedMessages =>
+            RecordedNotifications.GetMessages();
+
+        /// <summary>
+        /// Creates a new <see cref="FluentTestObserver{TPayload}"/> which subscribes to the supplied <see cref="IObservable{T}"/>
+        /// </summary>
+        /// <param name="subject">the <see cref="IObservable{T}"/> under test</param>
+        public FluentTestObserver(IObservable<TPayload> subject)
+        {
+            Subject = subject;
+            _observeScheduler.Schedule(() =>
+                Debug.WriteLine($"Observe scheduler thread id: {Thread.CurrentThread.ManagedThreadId}"));
+
+            _subscription = subject.ObserveOn(_observeScheduler).Subscribe(this);
+        }
+
+        /// <summary>
+        /// Clears the recorded notifications and messages as well as the recorded notifications stream buffer
+        /// </summary>
+        public void Clear() => _rollingReplaySubject.Clear();
+
+        /// <inheritdoc />
+        public void OnNext(TPayload value) =>
+            _rollingReplaySubject.OnNext(new Recorded<Notification<TPayload>>(_observeScheduler.Now.UtcTicks, Notification.CreateOnNext(value)));
+
+        /// <inheritdoc />
+        public void OnError(Exception exception) =>
+            _rollingReplaySubject.OnNext(new Recorded<Notification<TPayload>>(_observeScheduler.Now.UtcTicks, Notification.CreateOnError<TPayload>(exception)));
```

Comment:
Maybe OnError ?

Reply:
Nope, I basically copied that from the `ITestObserver` in *Microsoft.Reactive.Testing*... 

This "materializes" the tested subscription, see [introtorx](http://introtorx.com/Content/v1.0.10621.0/08_Transformation.html#MaterializeAndDematerialize)

Its equivalent to subscribing to the observable with 

```csharp
observable.Materialize().Select(n => new Recorded<Notification<TPayload>>(_observeScheduler.Now.UtcTicks, n).Subscribe(_rollingReplaySubject);
```

This totally makes sense, because you don't want the observer to "explode" on a failing subscription...